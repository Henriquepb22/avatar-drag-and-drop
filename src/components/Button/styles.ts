import styled, { css } from 'styled-components'

export const Wrapper = styled.button`
    ${({ theme }) => css`
        height: 3.2rem;
        border-radius: ${theme.radius.large};
        background: ${theme.colors.primary};
        color: ${theme.colors.white};
        font-family: ${theme.font.family};
        font-size: ${theme.font.sizes.small};
        font-weight: ${theme.font.medium};
        line-height: 180%;
        letter-spacing: -0.02em;
        border: 0;

        &:hover {
            cursor: pointer;
        }
    `}
`
