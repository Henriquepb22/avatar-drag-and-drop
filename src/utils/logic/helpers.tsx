const isImageFile = (file: File) => {
    const fileType = file.type
    const baseMimeType = fileType.split('/')[0]

    if (baseMimeType !== 'image') return false
    return true
}

export const getFileDataUrl = async (file: File): Promise<string> => {
    const reader = new FileReader()

    return new Promise((resolve, reject) => {
        const isImageType = isImageFile(file)

        if (isImageType) {
            reader.readAsDataURL(file)
            reader.onload = () => {
                if (reader.result) {
                    resolve(reader.result.toString())
                }
            }
        } else {
            reader.abort()
            reject(new Error('File is not an recognized image type'))
        }
    })
}

export const cropImage = async (
    imageSrc: string,
    zoomLevel: number
): Promise<string> => {
    const canvas = document.createElement('canvas')
    const context = canvas.getContext('2d')
    const image = new Image()

    return new Promise((resolve, reject) => {
        image.src = imageSrc

        image.onload = () => {
            if (context) {
                const needCompression =
                    Math.max(canvas.width, canvas.height) > 1000

                canvas.width = needCompression ? image.width / 2 : image.width
                canvas.height = needCompression
                    ? image.height / 2
                    : image.height

                context.drawImage(image, 0, 0, canvas.width, canvas.height)
                context.globalCompositeOperation = 'destination-in'
                context.beginPath()
                context.arc(
                    canvas.width * 0.5,
                    canvas.height * 0.5,
                    (Math.max(canvas.width, canvas.height) * 0.5) /
                        (zoomLevel * 2),
                    0,
                    2 * Math.PI
                )
                context.fill()
                context.globalCompositeOperation = 'source-over'

                resolve(canvas.toDataURL('image/png', 1))
            }
        }

        image.onerror = () => {
            reject(new Error('Something bad happened...'))
        }
    })
}
