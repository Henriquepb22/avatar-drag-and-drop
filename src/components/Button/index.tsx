import { ButtonHTMLAttributes } from 'react'
import * as S from './styles'

type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement>

const Button = ({ children, ...props }: ButtonProps) => {
    return <S.Wrapper {...props}>{children}</S.Wrapper>
}

export default Button
