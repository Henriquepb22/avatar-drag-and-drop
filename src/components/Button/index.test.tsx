import { renderWithTheme } from 'utils/tests/helpers'
import { screen } from '@testing-library/dom'

import Button from '.'

describe('<Button />', () => {
    it('should render button correctly', () => {
        const { container } = renderWithTheme(<Button>click here</Button>)

        expect(
            screen.getByRole('button', { name: /click here/i })
        ).toBeInTheDocument()
        expect(container.firstChild).toMatchSnapshot()
    })
})
