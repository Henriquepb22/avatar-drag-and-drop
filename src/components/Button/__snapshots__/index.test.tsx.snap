// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render button correctly 1`] = `
.c0 {
  height: 3.2rem;
  border-radius: 1.6rem;
  background: #3D485F;
  color: #FFFFFF;
  font-family: Inter,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.4rem;
  font-weight: 500;
  line-height: 180%;
  -webkit-letter-spacing: -0.02em;
  -moz-letter-spacing: -0.02em;
  -ms-letter-spacing: -0.02em;
  letter-spacing: -0.02em;
  border: 0;
}

.c0:hover {
  cursor: pointer;
}

<button
  class="c0"
>
  click here
</button>
`;
